fuzz/writeout.c

```diff
@@ -0,0 +1,216 @@
+        // +14 are arbitrary padding to avoid too many reallocs
+        char* nbuf = realloc(*buf, min_size + 14);
+        if (!nbuf) {
+            return false;
+        }
+        *buf = nbuf;
+        *buf_size = min_size + 14;
+    }
+
+    int ret = snprintf(*buf, *buf_size, "%ld:%02d:%02d.%02d",
+                       time * sign, m * sign, s * sign, cs * sign);
+    if (ret < 0 || ret >= *buf_size)
+        return false;
+    return true;
+}
+
+static int ssa2ass_align(int ssa_align)
+{
+    return ((ssa_align & 0xC) >> 2) * 3 + (ssa_align & 0x3);
+}
+
+static const char *tracktype_to_str(int track_type)
+{
+    return track_type == TRACK_TYPE_ASS ? "ASS"
+        : track_type == TRACK_TYPE_SSA ? "SSA"
+        : "other";
+}
+
+
+static void write_header(FILE *f, ASS_Track *track, const char* originalformat_name)
+{
+    // format version is normalised to ASS
+    fprintf(f, "[Script Info]\n");
+    fprintf(f, "; Original Format: %s\n", originalformat_name);
+    fprintf(f, "ScriptType: v4.00+\n");
+
+    #define HEADER_INT(name)  fprintf(f, "%s: %d\n",   #name, track->name)
+    #define HEADER_FLT(name)  fprintf(f, "%s: %.3f\n", #name, track->name)
+    #define HEADER_STR(name)  if (track->name) fprintf(f, "%s: %s\n",   #name, track->name)
+    #define HEADER_BOOL(name) fprintf(f, "%s: %s\n",   #name, track->name ? "yes" : "no")
+
+    HEADER_INT(PlayResX);
+    HEADER_INT(PlayResY);
+    HEADER_INT(LayoutResX);
+    HEADER_INT(LayoutResY);
+    HEADER_FLT(Timer);
+    HEADER_INT(WrapStyle);
+    HEADER_BOOL(ScaledBorderAndShadow);
+    HEADER_BOOL(Kerning);
+    HEADER_STR(Language);
+    if (track->YCbCrMatrix != YCBCR_DEFAULT) /* Or normalise this to TV.601? */
+        fprintf(f, "YCbCrMatrix: %s\n", ycbcr_to_str(track->YCbCrMatrix));
+
+    #undef HEADER_INT
+    #undef HEADER_FLT
+    #undef HEADER_STR
+    #undef HEADER_BOOL
+}
+
+static void write_styles(FILE *f, ASS_Track *track)
+{
+    fprintf(f, "\n[V4+ Styles]\n");
+    fprintf(f, "Format: "
+        "Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, "
+        "OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, "
+        "ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, "
+        "Alignment, MarginL, MarginR, MarginV, Encoding\n");
+
+    // don't print out builtin fallback
+    for (int i = 1; i < track->n_styles; ++i) {
+        ASS_Style *s = track->styles + i;
+        fprintf(f, "Style: "
+                   "%s,%s,%.3f,&H%08X,&H%08X,&H%08X,&H%08X,%d,%d,%d,%d,"
+                   "%.3f,%.3f,%.3f,%.3f,%d,%.3f,%.3f,%d,%d,%d,%d,%d\n",
+            s->Name, s->FontName, s->FontSize, s->PrimaryColour, s->SecondaryColour,
+            s->OutlineColour, s->BackColour, s->Bold, s->Italic, s->Underline, s->StrikeOut,
+            s->ScaleX, s->ScaleY, s->Spacing, s->Angle, s->BorderStyle, s->Outline, s->Shadow,
+            ssa2ass_align(s->Alignment), s->MarginL, s->MarginR, s->MarginV, s->Encoding
+            );
```

Comment:
I think per style guide `);` should be in the previous line.

Reply:
changed